### PR TITLE
Array auxiliary variable utilities

### PR DESCRIPTION
**Pull request by zachmprince**
This PR adds a couple new functionalities for array aux-variables:

- `ArrayParsedAux` which replicates `ParsedAux` for array variables.
- `SurrogateModelAuxKernel` which is similar to parsed aux, except it evaluates a surrogate model instead of a function expression.

Refs #20818

### CODE REVIEW

File: framework/include/auxkernels/ArrayParsedAux.h [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+  const unsigned int _n_array_vars;
+  /// Regular variable values
+  const std::vector<const VariableValue *> _vars;
+  /// Array variable values
+  const std::vector<const ArrayVariableValue *> _array_vars;
+
+  /// import coordinates and time
+  const bool _use_xyzt;
+
+  /// function parser object for the resudual and on-diagonal Jacobian
```
Comment (lindsayad) [inline framework/include/auxkernels/ArrayParsedAux.h:43]: seems not quite right for aux

File: test/tests/auxkernels/array_parsed_aux/array_parsed_aux.i [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+[AuxKernels]
+  [parsed_aux]
+    type = ArrayParsedAux
+    variable = parsed
+    expression = '(u^2 + v)*(x - const)*factor'
+    coupled_array_variables = 'u v'
+    coupled_variables = const
+    constant_names = 'factor'
+    constant_expressions = '3.14'
+    use_xyzt = true
```
Comment (lindsayad) [inline test/tests/auxkernels/array_parsed_aux/array_parsed_aux.i:70]: I'm wondering whether this should be the default?

**Review by lindsayad — DISMISSED** [review]
just reviewing the framework parts

File: test/tests/auxkernels/array_parsed_aux/array_parsed_aux.i (hunk 2 above)
Comment (zachmprince) [reply test/tests/auxkernels/array_parsed_aux/array_parsed_aux.i:70]: I'm willing to change, but should we stray from `ParsedAux` and `ParsedVectorAux` syntax?

Comment (lindsayad) [reply test/tests/auxkernels/array_parsed_aux/array_parsed_aux.i:70]: I suppose not in this PR. We can update them uniformly later if we want

**Review by grmnptr — COMMENTED** [review]
Super cool functionalities, Zach!

**Review by grmnptr — APPROVED** [review]

**Pull request merged**